lib/stan.js

```diff
@@ -271,116 +271,104 @@ Stan.prototype.initState = function () {
-      this.subCloseRequests = cr.getSubCloseRequests()
-      this.closeRequests = cr.getCloseRequests()
-
-      let unsubPingSub = true
-      if (cr.getProtocol() >= PROTOCOL_ONE) {
-        if (cr.getPingInterval() !== 0) {
-          unsubPingSub = false
-
-          this.pingRequests = cr.getPingRequests()
-          this.stanPingInterval = cr.getPingInterval() * 1000
-          this.stanMaxPingOut = cr.getPingMaxOut()
-
-          const ping = new proto.pb.Ping()
-          ping.setConnId(this.connId)
-          this.pingBytes = Buffer.from(ping.serializeBinary())
-
-          this.pingOut = 0
-          const that = this
-          this.pingTimer = setTimeout(function pingFun () {
-            that.pingOut++
-            if (that.pingOut > that.stanMaxPingOut) {
-              that.closeWithError('connection_lost', new Error(MAX_PINGS_EXCEEDED))
-              return
-            }
-            that.nc.publish(that.pingRequests, that.pingBytes, that.pingInbox)
-            that.pingTimer = setTimeout(pingFun, that.stanPingInterval)
-          }, this.stanPingInterval)
-        }
-      }
-      if (unsubPingSub) {
-        this.nc.unsubscribe(this.pingSubscription)
-        this.pingSubscription = null
+    const cr = proto.pb.ConnectResponse.deserializeBinary(Buffer.from(msg, 'binary'))
+    if (cr.getError() !== '') {
+      this.closeWithError('error', cr.getError())
+      return
+    }
+    this.pubPrefix = cr.getPubPrefix()
+    this.subRequests = cr.getSubRequests()
+    this.unsubRequests = cr.getUnsubRequests()
+    this.subCloseRequests = cr.getSubCloseRequests()
+    this.closeRequests = cr.getCloseRequests()
+
+    let unsubPingSub = true
+    if (cr.getProtocol() >= PROTOCOL_ONE) {
+      if (cr.getPingInterval() !== 0) {
+        unsubPingSub = false
+
+        this.pingRequests = cr.getPingRequests()
+        this.stanPingInterval = cr.getPingInterval() * 1000
+        this.stanMaxPingOut = cr.getPingMaxOut()
+
+        const ping = new proto.pb.Ping()
+        ping.setConnId(this.connId)
+        this.pingBytes = Buffer.from(ping.serializeBinary())
+
+        this.pingOut = 0
+        const that = this
+        this.pingTimer = setTimeout(function pingFun () {
+          that.pingOut++
+          if (that.pingOut > that.stanMaxPingOut) {
+            that.closeWithError('connection_lost', new Error(MAX_PINGS_EXCEEDED))
+            return
+          }
+          that.nc.publish(that.pingRequests, that.pingBytes, that.pingInbox)
+          that.pingTimer = setTimeout(pingFun, that.stanPingInterval)
+        }, this.stanPingInterval)
       }
+    }
+    if (unsubPingSub) {
+      this.nc.unsubscribe(this.pingSubscription)
+      this.pingSubscription = null
+    }
 
-      this.emit('connect', this)
-    })
+    this.emit('connect', this)
+  })
+}
+
+// @private
+Stan.prototype.setupHandlers = function () {
+  // if they gave us a connection, it could be connected, so the `connect` event may never fire.
```

Comment:
I would have done that in createConnection() since we already check if the connection is owned or not.